README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mvpd: An R Package for Multivariate Product Distributions

-   `[dpr]mvss`: mutivariate subgaussian stable distributions

<!-- badges: start -->
<!-- badges: end -->

The goal of `mvpd` is to use product distribution theory to allow the
numerical calculations of specific scale mixtures of the multivariate
normal distribution. The multivariate subgaussian stable distribution is
the product of the square root of a univariate positive stable
distribution and the multivariate normal distribution (see Nolan
(2013)).

<!-- ## Installation -->
<!-- You can install the development version of mvpd from [GitHub](https://github.com/) with: -->
<!-- ``` r -->
<!-- # install.packages("devtools") -->
<!-- devtools::install_github("swihart/mvpd") -->
<!-- ``` -->

## Example

Generate 1000 draws from a random bivariate subgaussian stable
distribution with alpha=1.71 and plot.

``` r
library(mvpd)
set.seed(10)
## basic example code
biv <- rmvss(n=1e3, alpha=1.71, Q=matrix(c(10,7.5,7.5,10),2))
head(biv)
#>            [,1]       [,2]
#> [1,] -0.2260798 -0.6168492
#> [2,] -6.1460819 -4.5603538
#> [3,]  1.4592466  1.6213040
#> [4,] -4.4159078 -2.9252448
#> [5,] -6.7106973 -3.8158068
#> [6,]  5.9107788  5.1332625
plot(biv); abline(h=0,v=0)
```

<img src="man/figures/README-example-1.png" width="100%" />


NEWS.md
# mvpd 0.0.4

* fixed roxygen token
* changed reverse depends from libstableR to libstable4u

# mvpd 0.0.3

* Improved help pages for fit_mvss and rmvss
* Added more tests

# mvpd 0.0.2

* Added multivariate subgaussian stable capabilities

# mvpd 0.0.1

* Added a `NEWS.md` file to track changes to the package.


R/mvpd.R
#' Multivariate Product Distributions
#'
#' The purpose of this package is to offer density, probability, and 
#' random variate generating (abbreviated as \[d/p/r\], respectively) 
#' functions for 
#' multivariate distributions that can be represented as a product distribution.
#' Specifically, the package will primarily focus on the product of a multivariate 
#' normal distribution and a univariate random variable. 
#' These product distributions are called Scale Mixtures of 
#' Multivariate Normal Distributions, 
#' and for particular choices of the univariate random variable distribution the 
#' resultant product distribution may be a family of interest.  For instance,
#' the square-root of a positive stable random variable multiplied by a 
#' multivariate normal distribution is the multivariate subgaussian stable
#' distribution. Product 
#' distribution theory is applied for implementing their computation.
#' 
#'
#' @section Multivariate subgaussian stable distributions:
#'
#' \code{\link{dmvss}} -- multivariate subgaussian stable distribution density
#' 
#' \code{\link{pmvss}} -- multivariate subgaussian stable distribution probabilities
#' 
#' \code{\link{rmvss}} -- multivariate subgaussian stable distribution random variates
#'
#' \code{\link{pmvss_mc}} -- Monte Carlo version of probabilities, using \code{rmvss}
#' 
#' \code{\link{fit_mvss}} -- Fit a multivariate subgaussian stable distribution (e.g. estimate parameters given data)
#' 
#' @docType package
#' @name mvpd
"_PACKAGE"
#> NULL


R/rmvt.R
#' Multivariate t Distribution Random Variates
#'
#'
#' Computes random vectors of the multivariate t distribution
#' distribution for arbitrary correlation matrices using inverse gamma variates
#' not sqrt(nu)/chi.
#'
#'
#' @param n number of observations
#' @param df default to 30.
#' @param delta location vector.
#' @param Q semi-positive definite
#'
#' @references
#'
#' Genz, A. and Bretz, F. (2002), Methods for the computation of multivariate
#' t-probabilities. \emph{Journal of Computational and Graphical Statistics},
#' \bold{11}, 950--971.
#'
#'
#' @keywords distribution
#' @importFrom stats    rnorm
#' @importFrom invgamma rinvgamma
#' @return Returns the \code{n} by \code{d} matrix containing multivariate t random variates
#' where \code{d=nrow(Q)}.
#' @examples
#' ## sample usage
#' rmvt(10, df=4, Q=diag(5))
#'
#' ## compare to rmvt from mvtnorm R-package
#' draw_mvtnorm <- mvtnorm::rmvt(2e4, sigma=diag(2), df = 35)
#' draw_mvpd    <-    mvpd::rmvt(2e4,     Q=diag(2), df = 35)
#'
#' mean(draw_mvtnorm[,1] < -1 & draw_mvtnorm[,2] < 3)
#' mean(draw_mvpd[,1] < -1 & draw_mvpd[,2] < 3)
#' 
#' rangex <- range(c(draw_mvpd[,1],draw_mvtnorm[,1]))
#' rangey <- range(c(draw_mvpd[,2],draw_mvtnorm[,2]))
#'
#' ## https://stackoverflow.com/a/53804499/2727349
#' opar <- par(mfrow=c(2,2))
#' plot(draw_mvtnorm, xlim=rangex, ylim=rangey); abline(h=0,v=0)
#' plot(draw_mvpd   , xlim=rangex, ylim=rangey); abline(h=0,v=0)
#'
#' hist(draw_mvtnorm, breaks = 100, ylim=c(0,2000), xlim=rangex)
#' hist(draw_mvpd    ,breaks = 100, ylim=c(0,2000), xlim=rangex)
#' par(opar)
#' @export
rmvt <- function(n, df=30, Q = NULL, delta=rep(0,d)
){
  d <- nrow(Q)
  
  A <- invgamma::rinvgamma(n, shape=df/2, rate=df/2)
  
  G <-   matrix(rnorm(n * d), n, d) %*% chol(Q)
  
  
  sqrt(A) * G + matrix(delta,nrow=n,ncol=length(delta),byrow=T)
  
  
}


R/dmvss.R
#' Multivariate Subgaussian Stable Density
#' 
#' 
#' Computes the the density function of the multivariate subgaussian stable
#' distribution for arbitrary alpha, shape matrices, and location vectors.
#' See Nolan (2013).
#' 
#' @param x vector of quantiles.
#' @param alpha default to 1 (Cauchy). Must be 0<alpha<2
#' @param delta location vector
#' @param Q Shape matrix.  See Nolan (2013).
#' @param outermost.int select which integration function to use for outermost
#' integral.  Default is "stats::integrate" and one can specify the following options 
#' with the `.si` suffix.  For diagonal Q, one can also specify "cubature::adaptIntegrate"
#' and use the `.ai` suffix options below (currently there is a bug for non-diagnoal Q).
#' @param spherical default is FALSE.  If true, use the spherical transformation.  
#' Results will be identical to spherical = FALSE but may be faster.
#' @param subdivisions.si the maximum number of subintervals.
#' The suffix \code{.si} indicates a \code{stats::integrate()} 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param rel.tol.si relative accuracy requested.  
#' The suffix \code{.si} indicates a \code{stats::integrate()} 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param abs.tol.si absolute accuracy requested. The suffix \code{.si} indicates a \code{stats::integrate()} 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param stop.on.error.si logical. If true (the default) an error stops the function. 
#' If false some errors will give a result with a warning in the message component.
#' The suffix \code{.si} indicates a \code{stats::integrate()} 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param tol.ai The maximum tolerance, default 1e-5.
#' The suffix \code{.ai} indicates a \code{cubature::adaptIntegrate} type 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param fDim.ai The dimension of the integrand, default 1, bears no
#' relation to the dimension of the hypercube
#' The suffix \code{.ai} indicates a \code{cubature::adaptIntegrate} type 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param maxEval.ai The maximum number of function evaluations needed,
#' default 0 implying no limit
#' The suffix \code{.ai} indicates a \code{cubature::adaptIntegrate} type 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param absError.ai The maximum absolute error tolerated
#' The suffix \code{.ai} indicates a \code{cubature::adaptIntegrate} type 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param doChecking.ai A flag to be thorough checking inputs to
#' C routines. A FALSE value results in approximately 9 percent speed
#' gain in our experiments. Your mileage will of course vary. Default
#' value is FALSE.
#' The suffix \code{.ai} indicates a \code{cubature::adaptIntegrate} type 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param which.stable defaults to "libstable4u", other option is "stabledist".  Indicates which package 
#' should provide the univariate stable distribution in this production distribution form of a univariate
#' stable and multivariate normal.
#' @return The object returned depends on what is selected for \code{outermost.int}.  In the case of the default, 
#' \code{stats::integrate}, the value is a list of class "integrate" with components:
#' \itemize{
#' \item{\code{value}}{ the final estimate of the integral.}
#' \item{\code{abs.error}}{ estimate of the modulus of the absolute error.}
#' \item{\code{subdivisions}}{ the number of subintervals produced in the subdivision process.}
#' \item{\code{message}}{ "OK" or a character string giving the error message.}
#' \item{\code{call}}{ the matched call.}
#' }
#' Note: The reported \code{abs.error} is likely an under-estimate as \code{integrate}
#' assumes the integrand was without error, 
#' which is not the case in this application.  
#' @references
#' 
#' Nolan, John P. "Multivariate elliptically contoured stable distributions: theory and estimation." Computational Statistics 28.5 (2013): 2067-2089.
#' 
#' @keywords distribution
#' @importFrom cubature adaptIntegrate
#' @importFrom matrixStats rowProds
#' @importFrom stabledist dstable
#' @importFrom libstable4u stable_pdf
#' @importFrom mvtnorm pmvnorm GenzBretz
#' @importFrom stats integrate dnorm
#' @examples
#' 
#' ## print("mvsubgaussPD (d=2, alpha=1.71):")
#' Q <- matrix(c(10,7.5,7.5,10),2)
#' mvpd::dmvss(x=c(0,1), alpha=1.71, Q=Q)
#' 
#' ## more accuracy = longer runtime
#' mvpd::dmvss(x=c(0,1),alpha=1.71, Q=Q, abs.tol=1e-8)
#' 
#' Q <- matrix(c(10,7.5,7.5,7.5,10,7.5,7.5,7.5,10),3)
#' ## print("mvsubgausPD (d=3, alpha=1.71):")
#' mvpd::dmvss(x=c(0,1,2), alpha=1.71, Q=Q)
#' mvpd::dmvss(x=c(0,1,2), alpha=1.71, Q=Q, spherical=TRUE)
#' 
#' ## How `delta` works: same as centering
#' X <- c(1,1,1)
#' Q <- matrix(c(10,7.5,7.5,7.5,10,7.5,7.5,7.5,10),3)
#' D <- c(0.75, 0.65, -0.35)
#' mvpd::dmvss(X-D, alpha=1.71, Q=Q)
#' mvpd::dmvss(X  , alpha=1.71, Q=Q, delta=D)
#' 
#' 
#' @export
dmvss <- function(x, alpha=1, Q = NULL, delta=rep(0,d), 
                          
                          outermost.int = c("stats::integrate","cubature::adaptIntegrate")[1],
                          
                          spherical = FALSE,
                          
                          subdivisions.si = 100L,
                          rel.tol.si = .Machine$double.eps^0.25, 
                          abs.tol.si = rel.tol.si,
                          stop.on.error.si = TRUE,
                          
                          tol.ai = 1e-05,
                          fDim.ai = 1,
                          maxEval.ai = 0,
                          absError.ai=0,
                          doChecking.ai=FALSE,
                          
                          which.stable=c("libstable4u", "stabledist")[1]
                          
                          ){
  
  f_A <- switch(which.stable,
                "libstable4u" =
                  function(x, alpha) 
                    libstable4u::stable_pdf(x,
                                           pars=c(
                                             alpha = alpha/2,
                                             beta = 1,
                                             sigma = 2*cos(pi*alpha/4)^(2/alpha),
                                             mu = 0),
                                           parametrization = 1L)
                ,
                "stabledist"=
                  function(x, alpha) 
                    stabledist::dstable(x,
                                        alpha = alpha/2,
                                        beta = 1,
                                        gamma = 2*cos(pi*alpha/4)^(2/alpha),
                                        delta = 0,
                                        pm = 1)
  )
  
  
  
  
  f_B <- switch(which.stable,
                "libstable4u" = 
                  function(x, alpha)
                    2*x*
                  libstable4u::stable_pdf(x^2,
                                         pars=c(
                                           alpha = alpha/2,
                                           beta = 1,
                                           sigma = 2*cos(pi*alpha/4)^(2/alpha),
                                           mu = 0),
                                         parametrization = 1L)
                ,
                "stabledist" = 
                  function(x, alpha)
                    2*x*
                  stabledist::dstable(x^2,
                                      alpha = alpha/2,
                                      beta = 1,
                                      gamma = 2*cos(pi*alpha/4)^(2/alpha),
                                      delta = 0,
                                      pm = 1)
  )  
    d <- length(x)
    x <- x - delta
    ifelse(!spherical,
           dens <- function(b,x,Qmat=Q,alp=alpha) f_B(b, alp) * 1/abs(b)^d * mvtnorm::dmvnorm(x/b, sigma=Qmat),
           dens <- function(b,x,Qmat=Q,alp=alpha){
             Linv <- solve(t(chol(Qmat)))
             detLinv <- det(Linv)
             f_B(b, alpha) * 1/abs(b)^d *detLinv *prod(dnorm(t(Linv %*% matrix(x/b,ncol=1))))
           }
    )
    
  switch(outermost.int,
         "cubature::adaptIntegrate"=
           adaptIntegrate_inf_limPD(dens,
                                    lowerLimit=0,
                                    upperLimit=Inf,
                                    x=x,
                                    Q=Q,
                                    tol.ai = tol.ai,
                                    fDim.ai = fDim.ai,
                                    maxEval.ai = maxEval.ai,
                                    absError.ai=absError.ai,
                                    doChecking.ai=doChecking.ai
                                    )
         
         ,
         "stats::integrate" =     
           integrate(Vectorize(dens,vectorize.args = "b"),
                     lower=0,
                     upper=Inf,
                     x=x,
                     Q=Q,
                     subdivisions = subdivisions.si,
                     rel.tol=rel.tol.si,
                     abs.tol=abs.tol.si,
                     stop.on.error=stop.on.error.si)
  )
  
  
  
  
  
}

# use ones in pmvsubgaussPD.R (now they are inside each [dp]mvsubgaussPD)
# f_A <- function(x, alpha) 
#   stabledist::dstable(x,
#                       alpha = alpha/2,
#                       beta = 1,
#                       gamma = 2*cos(pi*alpha/4)^(2/alpha),
#                       delta = 0,
#                       pm = 1)
# f_B <- function(x, alpha)
#   2*x*
#   stabledist::dstable(x^2,
#                       alpha = alpha/2,
#                       beta = 1,
#                       gamma = 2*cos(pi*alpha/4)^(2/alpha),
#                       delta = 0,
#                       pm = 1)


R/rmvss.R
#' Multivariate Subgaussian Stable Random Variates
#'
#'
#' Computes random vectors of the multivariate subgaussian stable
#' distribution for arbitrary alpha, shape matrices, and 
#' location vectors. See Nolan (2013).
#'
#'
#' @param n number of observations
#' @param alpha default to 1 (Cauchy). Must be 0<\code{alpha}<2
#' @param delta location vector.
#' @param Q Shape matrix.  See Nolan (2013).
#' @param which.stable defaults to \code{"libstable4u"}, other option is \code{"stabledist"}.  Indicates which package
#' should provide the univariate stable distribution in this production distribution form of a univariate
#' stable and multivariate normal.
#' @return Returns the \code{n} by \code{d} matrix containing multivariate subgaussian stable 
#' random variates where \code{d=nrow(Q)}.
#' @references
#' Nolan JP (2013), \emph{Multivariate elliptically contoured stable distributions:
#' theory and estimation}. Comput Stat (2013) 28:2067–2089
#' DOI 10.1007/s00180-013-0396-7
#' @keywords distribution
#' @importFrom matrixStats rowProds
#' @importFrom stabledist rstable
#' @importFrom libstable4u stable_rnd
#' @importFrom mvtnorm rmvnorm GenzBretz
#' @examples
#' ## generate 10 random variates of a bivariate mvss
#' rmvss(n=10, alpha=1.71, Q=matrix(c(10,7.5,7.5,10),2))
#' 
#' ## generate 10 random variates of a trivariate mvss
#' Q <- matrix(c(10,7.5,7.5,7.5,10,7.5,7.5,7.5,10),3)
#' rmvss(n=10, alpha=1.71, Q=Q)
#' 
#'
#' @export
rmvss <- function(n, alpha=1, Q = NULL, delta=rep(0,d),
                  which.stable=c("libstable4u", "stabledist")[1]
){
  d <- nrow(Q)

  A <- switch(which.stable,
              "libstable4u" =

                libstable4u::stable_rnd(n,
                                       pars=c(
                                         alpha = alpha/2,
                                         beta = 1,
                                         sigma = 2*cos(pi*alpha/4)^(2/alpha),
                                         mu = 0),
                                       parametrization = 1L)
              ,
              "stabledist"=

                stabledist::rstable(n,
                                    alpha = alpha/2,
                                    beta = 1,
                                    gamma = 2*cos(pi*alpha/4)^(2/alpha),
                                    delta = 0,
                                    pm = 1)
  )


  G <-   mvtnorm::rmvnorm(n, sigma=Q)


  sqrt(A) * G + matrix(delta,nrow=n,ncol=length(delta),byrow=T)


}


R/fit_mvss.R
#' Fit a Multivariate Subgaussian Distribution
#' 
#' 
#' Estimates the parameters (namely, alpha, shape matrix Q, and location vector) 
#' of the multivariate subgaussian
#' distribution for an input matrix X.
#' 
#' 
#' Using the protocols outlined in Nolan (2013), this function uses \code{libstable4u}'s univariate 
#' fit functions for each component.  
#' 
#' 
#' @param x a matrix for which the parameters for a \code{d}-dimensional multivariate 
#' subgaussian distribution will be estimated.  The number of columns will be \code{d}.
#' @return A list with parameters from the column-wise univariate fits and 
#' the multivariate alpha and shape matrix estimates (the \code{univ_deltas} are the \code{mult_deltas}):
#' \itemize{
#'   \item \code{univ_alphas} - the alphas from the column-wise univariate fits
#'   \item \code{univ_betas}  - the betas  from the column-wise univariate fits
#'   \item \code{univ_gammas} - the gammas from the column-wise univariate fits
#'   \item \code{univ_deltas} - the deltas from the column-wise univariate fits
#'   \item \code{mult_alpha}  - the mean(univ_alphas); equivalently the multivariate alpha estimate
#'   \item \code{mult_Q_raw}  - the multivariate shape matrix estimate (before applying \code{nearPD()})
#'   \item \code{mult_Q_posdef}   - the nearest positive definite multivariate shape matrix estimate, \code{nearPD(mult_Q_raw)}
#' }
#' @seealso  \code{Rfast::mvnorm.mle}, \code{alphastable::mfitstab.elliptical}
#' @references
#' Nolan JP (2013), \emph{Multivariate elliptically contoured stable distributions:
#' theory and estimation}. Comput Stat (2013) 28:2067–2089
#' DOI 10.1007/s00180-013-0396-7
#' 
#' @keywords distribution
#' @importFrom utils combn
#' @importFrom libstable4u stable_fit_mle2d
#' @importFrom Matrix nearPD
#' @examples
#' \donttest{
#' ## create a 4x4 shape matrix symMat
#' S <- matrix(rnorm(4*4, mean=2, sd=4),4); 
#' symMat <- as.matrix(Matrix::nearPD(0.5 * (S + t(S)))$mat)
#' symMat
#' ## generate 10,000 r.v.'s from 4-dimensional mvss
#' X <- mvpd::rmvss(1e4, alpha=1.5, Q=symMat, delta=c(1,2,3,4))
#' ## use fit_mvss to recover the parameters, compare to symMat
#' fmv <- mvpd::fit_mvss(X)
#' fmv
#' symMat
#' ## then use the fitted parameters to calculate a probability:
#' mvpd::pmvss(lower=rep(0,4),
#'             upper=rep(5,4),
#'             alpha=fmv$mult_alpha,
#'             Q=fmv$mult_Q_posdef,
#'             delta=fmv$univ_deltas,
#'             maxpts.pmvnorm = 25000*10)
#' }
#' 
#' @export
fit_mvss <- function(x){
  
  ## x comes in as an nxd matrix.  
  ## the task is to run univariate fit across all the columns
  component_results <- apply(x, 2, function(W){libstable4u::stable_fit_mle2d(W, parametrization = 1L)})
  
  ## now need to compute the shape matrix Q
  ## according to Nolan 2013, we subtract off the delta estimates
  xc <- x - matrix(component_results[4,],nrow=nrow(x),ncol=ncol(x),byrow=T)
  
  ## we know that the diagonal of shape matrix Q is just univariate gammas squared.
  ## to calculate the off-diagonal, we have to add up all pairwise combinations
  ## of components, fit a univariate stable distribution, take that estimated gamma
  ## and subtract off the diagonals and then divide by 2.  What's an efficient way
  ## to do this?
  ## https://stackoverflow.com/a/40139629/2727349
  all.pairs <- utils::combn(1:ncol(xc),2)
  xcp <- xc[, all.pairs[1,]] + xc[, all.pairs[2,]]
  
  if(NCOL(xcp)>1){
    ## now apply the same apply() we did to the original x to this xcp (c-centered p-pairwise sums)
    xcp_results <- apply(xcp, 2, function(W){libstable4u::stable_fit_mle2d(W, parametrization = 1L)})
    ## initialize Q_est
    Q_est <- diag(component_results[3,]^2)
    ## do a double assignment to get symmetry, loop over the pairwise combos
    for(i in 1:NCOL(all.pairs)){
      Q_est[all.pairs[1,i], all.pairs[2,i]] <- Q_est[all.pairs[2,i], all.pairs[1,i]] <- (xcp_results[3,i]^2 - Q_est[all.pairs[1,i], all.pairs[1,i]] - Q_est[all.pairs[2,i], all.pairs[2,i]])/2
    }
  }
  
  if(NCOL(xcp)==1){
    ## now apply the same apply() we did to the original x to this xcp (c-centered p-pairwise sums)
    xcp_results <- libstable4u::stable_fit_mle2d(xcp, parametrization = 1L)
    ## initialize Q_est
    Q_est <- diag(component_results[3,]^2)
    ## do a double assignment to get symmetry, loop over the pairwise combos
    for(i in 1:NCOL(all.pairs)){
      Q_est[all.pairs[1,i], all.pairs[2,i]] <- Q_est[all.pairs[2,i], all.pairs[1,i]] <- (xcp_results[3  ]^2 - Q_est[all.pairs[1,i], all.pairs[1,i]] - Q_est[all.pairs[2,i], all.pairs[2,i]])/2
    }
  }
  
  return(
    list(
      univ_alphas = component_results[1,],
      univ_betas  = component_results[2,],
      univ_gammas = component_results[3,],
      univ_deltas = component_results[4,],
      mult_alpha  = mean(component_results[1,]),
      mult_Q_raw  = Q_est,
      mult_Q_posdef   = as.matrix(Matrix::nearPD(Q_est)$mat)
    )
  )
  
}




R/pmvss_mc.R
#' Monte Carlo Multivariate Subgaussian Stable Distribution
#' 
#' 
#' Computes probabilities of the multivariate subgaussian stable
#' distribution for arbitrary limits, alpha, shape matrices, and
#' location vectors via Monte Carlo (thus the suffix `_mc`).
#' 
#' @param lower the vector of lower limits of length n.
#' @param upper the vector of upper limits of length n.
#' @param alpha default to 1 (Cauchy). Must be 0<alpha<2
#' @param delta location vector.
#' @param Q Shape matrix.  See Nolan (2013).
#' @param which.stable defaults to "libstable4u", other option is "stabledist".  Indicates which package 
#' should provide the univariate stable distribution in this production distribution form of a univariate
#' stable and multivariate normal.
#' @param n number of random vectors to be drawn for Monte Carlo calculation.
#' @return a number between 0 and 1, the estimated probability via Monte Carlo
#' @references
#' Nolan JP (2013), \emph{Multivariate elliptically contoured stable distributions:
#' theory and estimation}. Comput Stat (2013) 28:2067–2089
#' DOI 10.1007/s00180-013-0396-7
#' @examples
#' 
#' ## print("mvpd (d=2, alpha=1.71):")
#' U <- c(1,1)
#' L <- -U
#' Q <- matrix(c(10,7.5,7.5,10),2)
#' mvpd::pmvss_mc(L, U, alpha=1.71, Q=Q, n=1e3)
#' mvpd::pmvss   (L, U, alpha=1.71, Q=Q)
#'
#' ## more accuracy = longer runtime
#' mvpd::pmvss_mc(L, U, alpha=1.71, Q=Q, n=1e4)
#' 
#' U <- c(1,1,1)
#' L <- -U
#' Q <- matrix(c(10,7.5,7.5,7.5,10,7.5,7.5,7.5,10),3)
#' ## print("mvpd: (d=3, alpha=1.71):")
#' mvpd::pmvss_mc(L, U, alpha=1.71, Q=Q, n=1e3)
#' 
#' 
#' @export
pmvss_mc <- function(lower=rep(-Inf,d), upper=rep(Inf,d), alpha=1, Q = NULL, delta=rep(0,d), 
                          
                          which.stable=c("libstable4u", "stabledist")[1],
                          
                          n=NULL
                          
                          ){
  d <- nrow(Q)
  rmv <- mvpd::rmvss(n, alpha, Q, delta, which.stable)
  mc.result <- sum(rowSums(rmv >= lower & rmv <= upper) == nrow(Q))/n 
  mc.result
}




R/pmvss.R
#' Multivariate Subgaussian Stable Distribution
#' 
#' 
#' Computes the probabilities for the multivariate subgaussian stable
#' distribution for arbitrary limits, alpha, shape matrices, and 
#' location vectors.
#' See Nolan (2013).
#' 
#' @param lower the vector of lower limits of length n.
#' @param upper the vector of upper limits of length n.
#' @param alpha default to 1 (Cauchy). Must be 0<alpha<2
#' @param Q Shape matrix.  See Nolan (2013).
#' @param delta location vector.
#' @param maxpts.pmvnorm Defaults to 25000.  Passed to the F_G = pmvnorm() in the integrand of the outermost integral.
#' @param abseps.pmvnorm Defaults to 1e-3. Passed to the F_G = pmvnorm() in the integrand of the outermost integral.
#' @param outermost.int select which integration function to use for outermost
#' integral.  Default is "stats::integrate" and one can specify the following options 
#' with the `.si` suffix.  For diagonal Q, one can also specify "cubature::adaptIntegrate"
#' and use the `.ai` suffix options below (currently there is a bug for non-diagonal Q).
#' @param subdivisions.si the maximum number of subintervals.
#' The suffix \code{.si} indicates a \code{stats::integrate()} 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param rel.tol.si relative accuracy requested.  
#' The suffix \code{.si} indicates a \code{stats::integrate()} 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param abs.tol.si absolute accuracy requested. The suffix \code{.si} indicates a \code{stats::integrate()} 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param stop.on.error.si logical. If true (the default) an error stops the function. 
#' If false some errors will give a result with a warning in the message component.
#' The suffix \code{.si} indicates a \code{stats::integrate()} 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param tol.ai The maximum tolerance, default 1e-5.
#' The suffix \code{.ai} indicates a \code{cubature::adaptIntegrate} type 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param fDim.ai The dimension of the integrand, default 1, bears no
#' relation to the dimension of the hypercube
#' The suffix \code{.ai} indicates a \code{cubature::adaptIntegrate} type 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param maxEval.ai The maximum number of function evaluations needed,
#' default 0 implying no limit
#' The suffix \code{.ai} indicates a \code{cubature::adaptIntegrate} type 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param absError.ai The maximum absolute error tolerated
#' The suffix \code{.ai} indicates a \code{cubature::adaptIntegrate} type 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param doChecking.ai A flag to be thorough checking inputs to
#' C routines. A FALSE value results in approximately 9 percent speed
#' gain in our experiments. Your mileage will of course vary. Default
#' value is FALSE.
#' The suffix \code{.ai} indicates a \code{cubature::adaptIntegrate} type 
#' option for the outermost semi-infinite integral in the product distribution formulation.
#' @param which.stable defaults to "libstable4u", other option is "stabledist".  Indicates which package 
#' should provide the univariate stable distribution in this production distribution form of a univariate
#' stable and multivariate normal.
#' @return The object returned depends on what is selected for \code{outermost.int}.  In the case of the default, 
#' \code{stats::integrate}, the value is a list of class "integrate" with components:
#' \itemize{
#' \item{\code{value}}{ the final estimate of the integral.}
#' \item{\code{abs.error}}{ estimate of the modulus of the absolute error.}
#' \item{\code{subdivisions}}{ the number of subintervals produced in the subdivision process.}
#' \item{\code{message}}{ "OK" or a character string giving the error message.}
#' \item{\code{call}}{ the matched call.}
#' }
#' Note: The reported \code{abs.error} is likely an under-estimate as \code{integrate}
#' assumes the integrand was without error, 
#' which is not the case in this application.  
#' @references
#' Nolan JP (2013), \emph{Multivariate elliptically contoured stable distributions:
#' theory and estimation}. Comput Stat (2013) 28:2067–2089
#' DOI 10.1007/s00180-013-0396-7
#' 
#' @keywords distribution
#' @importFrom stabledist dstable
#' @importFrom libstable4u stable_pdf
#' @importFrom mvtnorm pmvnorm GenzBretz
#' @importFrom stats integrate
#' @examples
#' 
#' ## bivariate
#' U <- c(1,1)
#' L <- -U
#' Q <- matrix(c(10,7.5,7.5,10),2)
#' mvpd::pmvss(L, U, alpha=1.71, Q=Q)
#' 
#' ## trivariate
#' U <- c(1,1,1)
#' L <- -U
#' Q <- matrix(c(10,7.5,7.5,7.5,10,7.5,7.5,7.5,10),3)
#' mvpd::pmvss(L, U, alpha=1.71, Q=Q)
#' 
#' ## How `delta` works: same as centering
#' U <- c(1,1,1)
#' L <- -U
#' Q <- matrix(c(10,7.5,7.5,7.5,10,7.5,7.5,7.5,10),3)
#' D <- c(0.75, 0.65, -0.35)
#' mvpd::pmvss(L-D, U-D, alpha=1.71, Q=Q)
#' mvpd::pmvss(L  , U  , alpha=1.71, Q=Q, delta=D)
#' 
#' 
#' 
#' @export
pmvss <-
  function(lower=rep(-Inf,d), upper=rep(Inf,d), alpha=1, Q = NULL, delta=rep(0,d), 
           
           
           maxpts.pmvnorm = 25000,
           abseps.pmvnorm = 1e-3,
           
           outermost.int = c("stats::integrate","cubature::adaptIntegrate")[1],
           
           subdivisions.si = 100L,
           rel.tol.si = .Machine$double.eps^0.25, 
           abs.tol.si = rel.tol.si,
           stop.on.error.si = TRUE,
           
           tol.ai = 1e-05,
           fDim.ai = 1,
           maxEval.ai = 0,
           absError.ai=0,
           doChecking.ai=FALSE,
           
           which.stable=c("libstable4u", "stabledist")[1]
           
  ){
    
    
    f_A <- switch(which.stable,
                  "libstable4u" =
                    function(x, alpha) 
                      libstable4u::stable_pdf(x,
                                             pars=c(
                                               alpha = alpha/2,
                                               beta = 1,
                                               sigma = 2*cos(pi*alpha/4)^(2/alpha),
                                               mu = 0),
                                             parametrization = 1L)
                  ,
                  "stabledist"=
                    function(x, alpha) 
                      stabledist::dstable(x,
                                          alpha = alpha/2,
                                          beta = 1,
                                          gamma = 2*cos(pi*alpha/4)^(2/alpha),
                                          delta = 0,
                                          pm = 1)
    )
    
    
    
    
    f_B <- switch(which.stable,
                  "libstable4u" = 
                    function(x, alpha)
                      2*x*
                    libstable4u::stable_pdf(x^2,
                                           pars=c(
                                             alpha = alpha/2,
                                             beta = 1,
                                             sigma = 2*cos(pi*alpha/4)^(2/alpha),
                                             mu = 0),
                                           parametrization = 1L)
                  ,
                  "stabledist" = 
                    function(x, alpha)
                      2*x*
                    stabledist::dstable(x^2,
                                        alpha = alpha/2,
                                        beta = 1,
                                        gamma = 2*cos(pi*alpha/4)^(2/alpha),
                                        delta = 0,
                                        pm = 1)
    )
    
    
    d <- length(lower)
    lower <- lower - delta
    upper <- upper - delta
    integrand <- function(b,llp, ulp, Qmat=Q, alp=alpha){ 
      F_G <- mvtnorm::pmvnorm(lower=llp/b, 
                              upper=ulp/b, 
                              sigma=Qmat,
                              algorithm=mvtnorm::GenzBretz(maxpts = maxpts.pmvnorm, 
                                                           abseps=abseps.pmvnorm, 
                                                           releps = 0))
      # print(Sys.time())
      if( attr(F_G, "error") > abseps.pmvnorm){ print(paste0("WARNING: EARLY/ATYPICAL COMPLETION of numerical F_G:")) 
                                                print(paste0("abseps.pmvnorm for F_G is ", abseps.pmvnorm, ", ")) 
                                                print(paste0("and the actual error returned is more than that at ", attr(F_G, "error"))) 
                                                print(paste0("Consider increasing `maxpts.pmvnorm` AND/OR enlarging `abseps.pmvnorm`.")  )
                                                }
      
      f_B(b, alp) * F_G[1]
    }
    # integrand <- function(b,llp, ulp, Qmat=Q, alp=alpha) 
    #   f_B(b, alp) * mvtnorm::pmvnorm(lower=llp/b, upper=ulp/b, sigma=Qmat,
    #                                  algorithm=mvtnorm::GenzBretz(maxpts = maxpts.pmvnorm, abseps=abseps.pmvnorm, releps = 0))
    # 
    # integrand2 <- function(b,llp, ulp, Qmat=Q, alp=alpha) 
    #   f_B(b, alp) * mvtnorm::pmvnorm(llp/b, ulp/b)
    
    
    switch(outermost.int,
           "cubature::adaptIntegrate"=
             adaptIntegrate_inf_limPD(integrand,
                                      lowerLimit=0,
                                      upperLimit=Inf,
                                      llp=lower,
                                      ulp=upper,
                                      tol.ai = tol.ai,
                                      fDim.ai = fDim.ai,
                                      maxEval.ai = maxEval.ai,
                                      absError.ai=absError.ai,
                                      doChecking.ai=doChecking.ai
             )
           
           ,
           "stats::integrate" =     
             integrate(Vectorize(integrand,vectorize.args = "b"),
                       lower=0,
                       upper=Inf,
                       llp=lower,
                       ulp=upper,
                       subdivisions = subdivisions.si,
                       rel.tol=rel.tol.si,
                       abs.tol=abs.tol.si,
                       stop.on.error=stop.on.error.si)
    )
    
    
    
    
    
  }